substra/sdk/backends/local/backend.py

```diff
@@ -0,0 +1,523 @@
+            pkhash=fs.hash_file(spec.metrics),
+            name=spec.name,
+            owner=_BACKEND_ID,
+            test_dataset=test_dataset,
+            permissions={
+                "process": {
+                    "public": permissions.public,
+                    "authorized_ids": permissions.authorized_ids,
+                },
+            },
+            description=str(spec.description),
+            metrics=str(spec.metrics),
+            metadata=spec.metadata if spec.metadata else dict(),
+        )
+
+        # add objective to storage and update optionnally the associated dataset
+        objective = self._db.add(objective, exist_ok)
+        if spec.test_data_manager_key:
+            dataset.objective_key = objective.key
+
+        return objective
+
+    def __add_algo(self, model_class, spec, exist_ok, spec_options=None):
+        permissions = self.__compute_permissions(spec.permissions)
+        algo = model_class(
+            key=fs.hash_file(spec.file),
+            pkhash=fs.hash_file(spec.file),
+            name=spec.name,
+            owner=_BACKEND_ID,
+            permissions={
+                "process": {
+                    "public": permissions.public,
+                    "authorized_ids": permissions.authorized_ids,
+                },
+            },
+            file=str(spec.file),
+            description=str(spec.description),
+            metadata=spec.metadata if spec.metadata else dict(),
+        )
+        return self._db.add(algo, exist_ok)
+
+    def _add_algo(self, spec, exist_ok, spec_options=None):
+        return self.__add_algo(models.Algo, spec, exist_ok, spec_options=spec_options)
+
+    def _add_aggregate_algo(self, spec, exist_ok, spec_options=None):
+        return self.__add_algo(
+            models.AggregateAlgo, spec, exist_ok, spec_options=spec_options
+        )
+
+    def _add_composite_algo(self, spec, exist_ok, spec_options=None):
+        return self.__add_algo(
+            models.CompositeAlgo, spec, exist_ok, spec_options=spec_options
+        )
+
+    def __intersect_permissions(self, aggregate_set, element):
+        if not element.permissions.process.public:
+            if aggregate_set is None:
+                aggregate_set = set(element.permissions.process.authorized_ids)
+            else:
+                aggregate_set = aggregate_set & set(
+                    element.permissions.process.authorized_ids
+                )
+        return aggregate_set
+
+    def __add_compute_plan(
+        self,
+        traintuple_keys=None,
+        composite_traintuple_keys=None,
+        aggregatetuple_keys=None,
+        testtuple_keys=None,
+    ):
+        tuple_count = functools.reduce(
+            lambda x, y: x + (len(y) if y else 0),
+            [
+                traintuple_keys,
+                composite_traintuple_keys,
+                aggregatetuple_keys,
+                testtuple_keys,
+            ],
+            0,
+        )
```

Comment:
For the sake of clarity, I'd rather we use `tuple_count = len(traintuple_keys) + len(composite_traintuple_keys) + len(aggregatetuple_keys) + len(testtuple_keys)`.

Reply:
I agree but any of these can either be a list or None, I can do

```python
tuple_count = 0
if traintuple_keys is not None:
    tuple_count += len(traintuple_keys)
if ....
```

but I would like to avoid

```python
tuple_count = (len(traintuple_keys) if traintuple_keys is not None else 0) + ...
```

Reply:
Got it. What do you think about using sum() instead of reduce then (with a lambda to handle the None case)?